src/Engine/Graphics/BspRenderer.cpp

```diff
@@ -6,31 +6,46 @@
 
 #include "Library/Logger/Logger.h"
 
+// TODO(yoctozepto): we should not see it here
 BspRenderer *pBspRenderer = new BspRenderer();
 
-//----- (004B0EA8) --------------------------------------------------------
-void BspRenderer::AddFaceToRenderList_d3d(int node_id, int uFaceID) {
-    int pTransitionSector;  // ax@11
-    // int dotdist;                              // edx@15
 
+//----- (004B0EA8) --------------------------------------------------------
+void BspRenderer::AddFace(int node_id, int uFaceID) {
+    // TODO(yoctozepto): does the below happen?
     if (uFaceID >= pIndoor->pFaces.size())
-        return;
+        return;  // face out of range
+
     BLVFace *pFace = &pIndoor->pFaces[uFaceID];
 
     if (!pFace->isPortal()) {
+        // NOTE(yoctozepto): the below happens, e.g., on various stairs
+        if (pFace->Invisible())
+            return;  // nothing to render
+        // NOTE(yoctozepto): the below happens, e.g., on various stairs
+        if (pFace->uNumVertices < 3) {
+            return;  // nothing to render
+        }
+        // TODO(yoctozepto): does the below happen?
+        if (!pFace->GetTexture()) {
+            return;  // nothing to render
+        }
```

Comment:
Yes see #1560 - we should continue with drawing the face even if it doesnt have a texture

Reply:
Ack. Just note we didn't draw the face if it was missing the texture. This logic is here simply moved from OGL renderer (as it can be applied earlier). I can skip this and let it be drawn but I'm not sure what can of worms will open. :/ I will try breaking some textures and see.

Reply:
Passing these through might crash the game or not (and then render the face as water), depending on how other variables are set up (texlayer, texunit). Should we keep the skip or remove it and see what happens to users?

Reply:
I still think we should pass it through and handle it in OGL renderer. After all, if we didn't want to draw this poly it should be marked as invisible. In OGL we can pick a texture/ draw as a solid /other as a visual clue there is a problem. This should be added as an issue.

Reply:
Makes sense to me. Let's change it.

Reply:
Added TODOs and created https://github.com/OpenEnroth/OpenEnroth/issues/1877